### ISSUE TITLE
[inductor] Function returns wrong value

### ISSUE THREAD

**Issue by int3** [pytorch/pytorch#108963]
### 🐛 Describe the bug

In the code below, we expect `f()` to return a pair of identical values. However, it returns zero for the second value:

```python
import torch

@torch.compile()
def f():                                                                                                                                                                                                                              
    device = 'cpu'
    tensor = torch.zeros((1,), dtype=torch.double, device=device)                                                                                                                                                                     
    index = torch.tensor([0], dtype=torch.long, device=device)                                                                                                                                                                                        
    source = torch.rand((1,), dtype=torch.double, device=device)                                                                                                                                                                      
    out = tensor.index_add(0, index, source, alpha=2.) / 2                                                                                                                                                                            
    return source.item(), out.item()

torch.random.manual_seed(0)
print(f())  # prints (0.08809276670217514, 0.0)
```

* The issue goes away without `torch.compile()`
* The issue goes away if we replace `torch.rand((1,), ...)` with `torch.tensor([value], ...)`
* The issue goes away if we remove the `/ 2` (the 2nd value will now be double the first)
* The issue occurs even if we replace `/ 2` by `+ 1` (the 2nd value will now be 1.0)
* The issue happens regardless of whether we are running on cuda or on cpu

Looking at the generated cpp, it looks like we are simply dividing the wrong variable:

```cpp
extern "C" void kernel(const long* in_ptr0,
                       double* out_ptr0,
                       double* out_ptr2,
                       double* out_ptr3)
{
    {
        auto tmp0 = static_cast<double>(0.0);
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        auto tmp6 = out_ptr0[static_cast<long>(0L)];
        auto tmp1 = static_cast<int>(0);
        auto tmp2 = normalized_rand_cpu(tmp0, tmp1);
        auto tmp3 = static_cast<double>(tmp2);
        auto tmp4 = static_cast<double>(2.0);
        auto tmp5 = decltype(tmp3)(tmp3 * tmp4);
        auto tmp7 = tmp6 / tmp4; // this should probably be tmp5 / tmp4
        out_ptr2[static_cast<long>(0L)] = tmp3;
        atomic_add(&out_ptr0[static_cast<long>(0L)], tmp5);
        out_ptr3[static_cast<long>(0L)] = tmp7;
    }
}
```

The traced graph seems fine though, at least to my untrained eye:

```python
def forward(self):
    # File: /home/jezng/tmp/test.py:6, code: tensor = torch.zeros((1,), dtype=torch.double, device=device)
    full: f64[1] = torch.ops.aten.full.default([1], 0, dtype = torch.float64, layout = torch.strided, device = device(type='cpu'), pin_memory = False)

    # File: /home/jezng/tmp/test.py:7, code: index = torch.tensor([0], dtype=torch.long, device=device)
    _tensor_constant0 = self._tensor_constant0
    lift_fresh_copy: i64[1] = torch.ops.aten.lift_fresh_copy.default(_tensor_constant0);  _tensor_constant0 = None

    # File: /home/jezng/tmp/test.py:8, code: source = torch.rand((1,), dtype=torch.double, device=device)
    rand: f64[1] = torch.ops.aten.rand.default([1], dtype = torch.float64, device = device(type='cpu'), pin_memory = False)

    # File: /home/jezng/tmp/test.py:9, code: out = tensor.index_add(0, index, source, alpha=2.) / 2
    mul: f64[1] = torch.ops.aten.mul.Tensor(rand, 2.0)
    index_put: f64[1] = torch.ops.aten.index_put.default(full, [lift_fresh_copy], mul, True);  full = lift_fresh_copy = mul = None
    div: f64[1] = torch.ops.aten.div.Tensor(index_put, 2);  index_put = None
    return (rand, div)
```

Happy to dig into it more, but I'll need some pointers on where to look.

### Versions

```
PyTorch version: 2.2.0a0+gitebed490
Is debug build: False
CUDA used to build PyTorch: 12.0
ROCM used to build PyTorch: N/A

OS: CentOS Stream 9 (x86_64)
GCC version: (GCC) 11.4.1 20230605 (Red Hat 11.4.1-2)
Clang version: Could not collect
CMake version: version 3.26.4
Libc version: glibc-2.34

Python version: 3.11.4 (main, Jul  5 2023, 13:45:01) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.12.0-0_fbk16_zion_7661_geb00762ce6d2-x86_64-with-glibc2.34
Is CUDA available: True
CUDA runtime version: 12.0.140
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration:
GPU 0: NVIDIA PG509-210
GPU 1: NVIDIA PG509-210
GPU 2: NVIDIA PG509-210
GPU 3: NVIDIA PG509-210
GPU 4: NVIDIA PG509-210
GPU 5: NVIDIA PG509-210
GPU 6: NVIDIA PG509-210
GPU 7: NVIDIA PG509-210

Nvidia driver version: 525.105.17
cuDNN version: Probably one of the following:
/usr/lib64/libcudnn.so.8.9.4
/usr/lib64/libcudnn_adv_infer.so.8.9.4
/usr/lib64/libcudnn_adv_train.so.8.9.4
/usr/lib64/libcudnn_cnn_infer.so.8.9.4
/usr/lib64/libcudnn_cnn_train.so.8.9.4
/usr/lib64/libcudnn_ops_infer.so.8.9.4
/usr/lib64/libcudnn_ops_train.so.8.9.4
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                    x86_64
CPU op-mode(s):                  32-bit, 64-bit
Address sizes:                   46 bits physical, 48 bits virtual
Byte Order:                      Little Endian
CPU(s):                          192
On-line CPU(s) list:             0-191
Vendor ID:                       GenuineIntel
Model name:                      Intel(R) Xeon(R) Platinum 8339HC CPU @ 1.80GHz
CPU family:                      6
Model:                           85
Thread(s) per core:              2
Core(s) per socket:              24
Socket(s):                       4
Stepping:                        11
Frequency boost:                 enabled
CPU max MHz:                     1801.0000
CPU min MHz:                     800.0000
BogoMIPS:                        3600.00
Flags:                           fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush dts acpi mmx fxsr sse sse2 ss ht tm pbe syscall nx pdpe1gb rdtscp lm constant_tsc art arch_perfmon pebs bts rep_good nopl xtopology nonstop_tsc cpuid aperfmperf pni pclmulqdq dtes64 monitor ds_cpl vmx smx est tm2 ssse3 sdbg fma cx16 xtpr pdcm pcid dca sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand lahf_lm abm 3dnowprefetch cpuid_fault epb cat_l3 cdp_l3 invpcid_single intel_ppin ssbd mba ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid cqm mpx rdt_a avx512f avx512dq rdseed adx smap clflushopt clwb intel_pt avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves cqm_llc cqm_occup_llc cqm_mbm_total cqm_mbm_local avx512_bf16 dtherm ida arat pln pts hwp hwp_act_window hwp_epp hwp_pkg_req pku ospke avx512_vnni md_clear flush_l1d arch_capabilities
Virtualization:                  VT-x
L1d cache:                       3 MiB (96 instances)
L1i cache:                       3 MiB (96 instances)
L2 cache:                        96 MiB (96 instances)
L3 cache:                        132 MiB (4 instances)
NUMA node(s):                    4
NUMA node0 CPU(s):               0-23,96-119
NUMA node1 CPU(s):               24-47,120-143
NUMA node2 CPU(s):               48-71,144-167
NUMA node3 CPU(s):               72-95,168-191
Vulnerability Itlb multihit:     Not affected
Vulnerability L1tf:              Not affected
Vulnerability Mds:               Not affected
Vulnerability Meltdown:          Not affected
Vulnerability Spec store bypass: Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:        Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:        Mitigation; Enhanced IBRS, IBPB conditional, RSB filling
Vulnerability Srbds:             Not affected
Vulnerability Tsx async abort:   Not affected

Versions of relevant libraries:
[pip3] bert-pytorch==0.0.1a4
[pip3] flake8==6.0.0
[pip3] flake8-bugbear==23.3.23
[pip3] flake8-comprehensions==3.12.0
[pip3] flake8-executable==2.1.3
[pip3] flake8-logging-format==0.9.0
[pip3] flake8-pyi==23.3.1
[pip3] flake8-simplify==0.19.3
[pip3] functorch==1.14.0a0+b71aa0b
[pip3] mypy==1.4.1
[pip3] mypy-extensions==1.0.0
[pip3] numpy==1.24.3
[pip3] pytorch-transformers==1.2.0
[pip3] pytorch-triton==2.1.0+9e3e10c5ed
[pip3] torch==2.2.0a0+git742b5a0
[pip3] torch_geometric==2.4.0
[pip3] torch-struct==0.5
[pip3] torchaudio==2.1.0a0+61cbf79
[pip3] torchdata==0.7.0a0+757c032
[pip3] torchdata==0.7.0a0+757c032
[pip3] torchmetrics==1.0.1
[pip3] torchmultimodal==0.1.0b0
[pip3] torchrec-nightly==2023.7.26
[pip3] torchtext==0.16.0a0+b5e2a1b
[pip3] torchvision==0.16.0a0+3966f95
[pip3] triton==2.1.0
[conda] bert-pytorch              0.0.1a4                   dev_0    <develop>
[conda] blas                      1.0                         mkl
[conda] functorch                 1.14.0a0+b71aa0b          pypi_0    pypi
[conda] magma-cuda116             2.6.1                         1    pytorch
[conda] mkl                       2023.1.0         h6d00ec8_46342
[conda] mkl-include               2023.1.0         h06a4308_46342
[conda] mkl-service               2.4.0           py311h5eee18b_1
[conda] mkl_fft                   1.3.6           py311ha02d727_1
[conda] mkl_random                1.2.2           py311ha02d727_1
[conda] numpy                     1.24.3          py311h08b1b3b_1
[conda] numpy-base                1.24.3          py311hf175353_1
[conda] pytorch-transformers      1.2.0                    pypi_0    pypi
[conda] pytorch-triton            2.1.0+9e3e10c5ed          pypi_0    pypi
[conda] torch                     2.2.0a0+git742b5a0           dev_0    <develop>
[conda] torch-geometric           2.4.0                    pypi_0    pypi
[conda] torch-struct              0.5                      pypi_0    pypi
[conda] torchaudio                2.1.0a0+61cbf79           dev_0    <develop>
[conda] torchdata                 0.7.0a0+757c032          pypi_0    pypi
[conda] torchmetrics              1.0.1                    pypi_0    pypi
[conda] torchmultimodal           0.1.0b0                  pypi_0    pypi
[conda] torchrec-nightly          2023.7.26                pypi_0    pypi
[conda] torchtext                 0.16.0a0+b5e2a1b           dev_0    <develop>
[conda] torchvision               0.16.0a0+3966f95           dev_0    <develop>
[conda] triton                    2.1.0                    pypi_0    pypi
```

cc @ezyang @gchanan @zou3519 @kadeng @msaroufim @wconstab @bdhirsh @anijain2305

**Comment by davidradl** [comment]
@int3 one way to test your theory (which seems reasonable to me) is to run the test python with debug on, and find the generated file from the debug output. You can then edit the C code and rerun the test; it will pick up the cached file that you have edited.

You may be interested in [https://github.com/pytorch/pytorch/pull/108513](https://github.com/pytorch/pytorch/pull/108513) and it's reported issue. These give 2 other ways of testing this sort of scenario:
1) the submitter compares the compiler and non compiler input to check they are the same
2) the pull request uses the standard pytorch framework to test compiler and non compiler code work identically.  

I would look at [https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/cpp.py](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/cpp.py) and [https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/triton.py](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/triton.py). Do a find for div in these files and you will see the existing division specific logic for cpu and triton.  fyi there is also a common file [https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/common.py](https://github.com/pytorch/pytorch/blob/main/torch/_inductor/codegen/common.py ) which cpu and triton share.

**Comment by leslie-fang-intel** [comment]
Update some of my findings, here is the generated code which produces wrong result:
```
cpp_fused_div_index_put_mul_0 = async_compile.cpp('''
#include "/tmp/torchinductor_root/ib/cibrnuq56cxamjj4krp4zpjvsirbmlolpbnmomodzyd46huzhdw7.h"
extern "C" void kernel(const double* in_ptr0,
                       const long* in_ptr1,
                       const double* in_ptr2,
                       double* out_ptr0,
                       double* out_ptr1)
{
    {
        auto tmp0 = in_ptr0[static_cast<long>(0L)];
        out_ptr0[static_cast<long>(0L)] = tmp0;
    }
    {
        auto tmp0 = in_ptr1[static_cast<long>(0L)];
        auto tmp1 = in_ptr2[static_cast<long>(0L)];
        auto tmp4 = out_ptr0[static_cast<long>(0L)];
        auto tmp2 = static_cast<double>(2.0);
        auto tmp3 = decltype(tmp1)(tmp1 * tmp2);
        auto tmp5 = tmp4 / tmp2;
        atomic_add(&out_ptr0[static_cast<long>(0L)], tmp3);
        out_ptr1[static_cast<long>(0L)] = tmp5;
    }
}
''')
```

I think the key error is that 

- We get `tmp4` from `out_ptr0` at very beginning.
- However, `atomic_add(&out_ptr0[static_cast<long>(0L)], tmp3);` will update the `out_ptr0`. So, we actually get the wrong tmp4 value. If we get `tmp4` from `out_ptr0` after `atomic_add`. The result looks good.

I will further dig it deeper. cc @jgong5, not sure any special logic to deal with `atomic_add`

**Comment by jgong5** [comment]
> I will further dig it deeper. cc @jgong5, not sure any special logic to deal with `atomic_add`

It is not related to the cpp codegen (or how atomic_add is handled). Seems it is related to the fusion logic in the inductor scheduler when there are dependencies with mutations. The issue won't occur if you manually disable fusion (comment out `self.fuse_nodes()` call in the scheduler.py). In this example, the `index_add` is decompsed into `index_put` and then lowered with `index_put_impl_` which does mutation into the buffer that passed to the `div`. The `div` depends on a buffer that is mutated by `ir.Scatter` (i.e., atomic_add) generated by `index_put_impl_` while if these operations are fused the `div` would be computed before the `atomic_add` which caused problems. The inductor ir_pre_fusion.txt is shown below where buf4 is the result of `atomic_add` and buf5 is the result of `div`. buf4 mutates buf3 which is loaded by buf5. We need to make sure buf4 is scheduled before buf5 by disabling such fusions.
```
buf4: SchedulerNode(ComputedBuffer)
buf4.writes = [MemoryDep('buf4', 0, {})]
buf4.unmet_dependencies = [MemoryDep('buf2', 0, {}), StarDep(name='buf3')]
buf4.met_dependencies = []
buf4.users = [NodeUser(node=SchedulerNode(name='buf5'), can_inplace=True, is_weak=False)]
buf4.group.device = cpu
buf4.group.iteration = ((), ())
buf4.sizes = ([], [])
buf4.mutations = ['buf3']
class buf4_loop_body:
    var_ranges = {}
    index0 = 0
    def body(self, ops):
        get_index = self.get_index('index0')
        load = ops.load('buf2', get_index)
        constant = ops.constant(2.0, torch.float64)
        mul = ops.mul(load, constant)
        get_index_1 = self.get_index('index0')
        store = ops.store('buf4', get_index_1, mul, 'atomic_add')
        return store

buf5: SchedulerNode(ComputedBuffer)
buf5.writes = [MemoryDep('buf5', 0, {})]
buf5.unmet_dependencies = [MemoryDep('buf4', 0, {})]
buf5.met_dependencies = []
buf5.users = [NodeUser(node=OUTPUT, can_inplace=False, is_weak=False)]
buf5.group.device = cpu
buf5.group.iteration = ((), ())
buf5.sizes = ([], [])
class buf5_loop_body:
    var_ranges = {}
    index0 = 0
    def body(self, ops):
        get_index = self.get_index('index0')
        load = ops.load('buf3', get_index)
        constant = ops.constant(2.0, torch.float64)
        div = ops.div(load, constant)
        get_index_1 = self.get_index('index0')
        store = ops.store('buf5', get_index_1, div, None)
        return store
```

**Comment by int3** [comment]
Fixed by 9e2b07ac9d189ff05c37a5d1683e239fd36dfe5b.
